torchsynth/synth.py

```diff
@@ -163,9 +165,178 @@ def _forward(self) -> T:
         pitch_envelope = self.pitch_adsr.forward(note_on_duration)
         amp_envelope = self.amp_adsr.forward(note_on_duration)
 
-        audio_out = self.vco_1.forward(midi_f0, pitch_envelope)
+        vco_1_out = self.vco_1.forward(midi_f0, pitch_envelope)
         vco_2_out = self.vco_2.forward(midi_f0, pitch_envelope)
-        audio_out = util.crossfade2D(audio_out, vco_2_out, self.vco_ratio.p("ratio"))
+        audio_out = util.crossfade2D(vco_1_out, vco_2_out, self.vco_ratio.p("ratio"))
 
         audio_out = self.noise.forward(audio_out)
         return self.vca.forward(amp_envelope, audio_out)
+
+
+class FmOperator(AbstractSynth):
+    """
+    An FM operator - a single unit containing a sine wave oscillator that can receive
+    an audio rate frequency modulation signal and contains an amplitude envelope
+    """
+
+    def __init__(self, synthglobals: SynthGlobals, *args, **kwargs):
+        AbstractSynth.__init__(self, synthglobals=synthglobals, *args, **kwargs)
+
+        # Register all modules as children
+        self.add_synth_modules(
+            [
+                ("osc", FmVCO(synthglobals)),
+                ("env", ADSR(synthglobals)),
+                ("amp", VCA(synthglobals)),
+            ]
+        )
+
+    def _forward(self, midi_f0: T, duration: T, modulation: Signal) -> Signal:
+        output = self.osc(midi_f0, modulation)
+        env = self.env(duration)
+        return self.amp(env, output)
+
+
+class FmAlgorithmKnob(SynthModule):
+    """
+    A knob for transitioning between FM algorithms
+    """
+
+    parameter_ranges: List[ModuleParameterRange] = [
+        ModuleParameterRange(
+            0.0,
+            10.0,
+            curve=1.0,
+            name="algorithm",
+            description="Algorithm mapping",
+        ),
+    ]
+
+
+class FmSynth(AbstractSynth):
+    """
+    A four operator FM synth. Has 10 different algorithms from Ableton's Operator
+    instrument and can smoothly transition between them.
+
+    TODO: Add an LFO and maybe a pitch envelope? Try to implement DX7 style envs?
+    """
+
+    def __init__(self, synthglobals: SynthGlobals, *args, **kwargs):
+        AbstractSynth.__init__(self, synthglobals=synthglobals, *args, **kwargs)
+
+        # Register children SynthModules
+        self.add_synth_modules(
+            [
+                ("keyboard", MonophonicKeyboard(synthglobals)),
+                ("algorithm", FmAlgorithmKnob(synthglobals)),
+            ]
+        )
+
+        # Add the operators
+        self.op1 = FmOperator(synthglobals)
+        self.op2 = FmOperator(synthglobals)
+        self.op3 = FmOperator(synthglobals)
+        self.op4 = FmOperator(synthglobals)
+
+        # Algorithm layouts - 11 different layouts from Ableton's Operator
+
```

Comment:
This is super confusing all these floats, can you give a little more detail here